compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2596,7 +2596,14 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         linkConstructorParams(sym, tparamSyms, rhsCtx)
 
     if sym.isInlineMethod then rhsCtx.addMode(Mode.InlineableBody)
-    if sym.is(ExtensionMethod) then rhsCtx.addMode(Mode.InExtensionMethod)
+    if sym.is(ExtensionMethod) then
+      if ctx.phase.isTyper then
+        val ValDef(_, paramTpt, _) = termParamssIn(paramss1).head.head
+        if !paramTpt.symbol.denot.isAliasType && !paramTpt.symbol.denot.isOpaqueAlias then
+          val other = paramTpt.tpe.nonPrivateMember(name)
+          if other.exists && sym.denot.info.resultType.matches(other.info) then
```

Comment:
There's a few situations where this check wouldn't be enough:

1. The method defined in the class is overloaded (`other.exists` will return false).

2. The parameters in the extension definition are subtypes of the one in the primary definition:
```scala
class C { def foo(x: Any) = 0 }
extension (c: C) def foo(x: Int) = 1 // uncallable
```

2. Only the first (if any) parameter list matches:
```scala
class C { def foo(x: Int)(y: String) = 0 }
extension (c: C) def foo(x: Int)(y: Int) = 1 // uncallable
```
```scala
class C { def foo: Int = 0 }
extension (c: C) def foo: String = "" // uncallable
```

4. The primary definition might have a type parameter or implicit/using clause:
```scala
class C { def foo(using Int) = 0 }
extension (c: C) def foo = 1 // uncallable
```

5. The extension definition might have a type parameter or implicit/using clause in its prefix:
```scala
class C { def foo = 0 }
extension (using Int)(c: C) def foo = 1 // uncallable
```

6. Borderline (which we may or may not want to warn about):  the extension definition might have a type parameter or implicit/using clause in its suffix:
```scala
class C { def foo = 0 }
extension (c: C) def foo(using Int) = 1 // only callable by explicitly passing `(using ...)`
```

I think this is solvable by doing something like:

```suggestion
          def firstExplicitParamTypes(tp: Type) = stripImplicit(tp.stripPoly, wildcardOnly = true).firstParamTypes
          val extFirst = firstExplicitParamTypes(sym.denot.info)
          val matching =
            paramTpt.tpe.nonPrivateMember(name)
              .filterWithPredicate: clsMeth =>
                val clsFirst = firstExplicitParamTypes(clsMeth.info)
                extFirst.lazyZip(clsFirst).forall(_ frozen_<:< _)
              .nonEmpty
```
But I haven't actually tried it.

Reply:
Thanks! I was bored and in need of stimulation. If I'm able to work through the cases, I'm sorry in advance that I will request your re-review.

Reply:
case 7 is case 6 but where `c.foo(using x)` typechecks with native member returning `{ def apply(using X) }`. I had to refresh my knowledge, and the reference is not current or precise (as it mentions only when m is missing, not when m is present but expression doesn't typecheck). I'll experiment now with your code suggestion.